crates/iota-build-cache-server/src/server.rs

```diff
@@ -0,0 +1,388 @@
+            error!("Failed to resolve commit '{commit_ref}': {e}");
+            Ok(Response::builder()
+                .status(StatusCode::BAD_REQUEST)
+                .header(CONTENT_TYPE, "text/plain")
+                .body(Full::new(Bytes::from(format!(
+                    "Invalid commit/branch/tag: {e}"
+                ))))
+                .unwrap())
+        }
+    }
+}
+
+/// Handle build requests
+async fn handle_build_request(
+    req: Request<Incoming>,
+    cache: Arc<BuildCache>,
+) -> Result<Response<Full<Bytes>>, Infallible> {
+    let body = req.collect().await.unwrap().to_bytes();
+
+    // Parse the request body as BuildRequest
+    match serde_json::from_slice::<BuildRequest>(&body) {
+        // Resolve branch/tag/commit to actual commit hash
+        Ok(build_request) => match cache.resolve_commit(build_request.commit.as_str()).await {
+            Ok(resolved_commit) => {
+                // Start the build
+                match cache
+                    .start_build(
+                        &resolved_commit,
+                        &build_request.cpu_target,
+                        &build_request.binaries,
+                    )
+                    .await
+                {
+                    Ok(build_response) => {
+                        let json = serde_json::to_string(&build_response).unwrap();
+                        Ok(Response::builder()
+                            .status(StatusCode::ACCEPTED)
+                            .header(CONTENT_TYPE, "application/json")
+                            .body(Full::new(Bytes::from(json)))
+                            .unwrap())
+                    }
+                    Err(e) => {
+                        error!("Failed to start build: {e}");
+                        Ok(Response::builder()
+                            .status(StatusCode::INTERNAL_SERVER_ERROR)
+                            .header(CONTENT_TYPE, "text/plain")
+                            .body(Full::new(Bytes::from(format!(
+                                "Failed to start build: {e}",
+                            ))))
+                            .unwrap())
+                    }
+                }
+            }
+            Err(e) => {
+                error!("Failed to resolve commit '{}': {e}", build_request.commit);
+                Ok(Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .header(CONTENT_TYPE, "text/plain")
+                    .body(Full::new(Bytes::from(format!(
+                        "Invalid commit/branch/tag: {e}"
+                    ))))
+                    .unwrap())
+            }
+        },
+        Err(e) => {
+            warn!("Invalid build request: {e}");
+            Ok(Response::builder()
+                .status(StatusCode::BAD_REQUEST)
+                .header(CONTENT_TYPE, "text/plain")
+                .body(Full::new(Bytes::from(format!("Invalid request: {e}"))))
+                .unwrap())
```

Comment:
There's a lot of repetition for these responses, might want to refactor with fns like these
```Rust
fn bad_request(msg: impl Into<String>) -> Response {
    (StatusCode::BAD_REQUEST, [(CONTENT_TYPE, "text/plain")], msg.into()).into_response()
}

fn internal_error(msg: impl Into<String>) -> Response {
    let msg = msg.into();
    error!("{msg}");
    (StatusCode::INTERNAL_SERVER_ERROR, [(CONTENT_TYPE, "text/plain")], msg).into_response()
}
```

Reply:
Fixed, thanks for the suggestion! Makes sense :D